### PR TITLE
Improve PriorityQueue performance with large priority_levels

### PR DESCRIPTION
**Pull request by QMassoz**
Modify PriorityQueue to dynamically create/remove PolicyQueue items instead of instantiating them in the constructor. User-given PolicyQueue (defined in priority_queue_policy) are instantiated in constructor and are kept instantiated. Thus only default PolicyQueue are affected by this PR.

This PR results in significant performance boost when priority_levels takes a large value (above ~100k). Next step would be to have the priority_ variable of InferenceRequest to be expressed in uint64_t instead of uint32_t.

### CODE REVIEW

**Comment by QMassoz** [comment]
Hello, how do we request reviews?

File: src/scheduler_utils.cc
Comment (dyastremsky) [inline src/scheduler_utils.cc:281]: For consistency with the other code in the core repo, can we use the shorter form a for loop here? E.g. for (const auto& queue_policy : queue_policy_map)

File: src/scheduler_utils.cc
Comment (dyastremsky) [inline src/scheduler_utils.cc:320]: Shorter for loop for consistency.

File: src/scheduler_utils.cc
Comment (QMassoz) [reply src/scheduler_utils.cc]: Done, replaced it with a range-based for loop

File: src/scheduler_utils.cc
Comment (QMassoz) [reply src/scheduler_utils.cc:319]: A range-based for loop (e.g. for (auto& queue : queues_)) will range from begin to end and not benefit from the (small) optimization using it_start (found in logarithmic time using front_priority_level_).

A suggestion for a shorter for loop:
auto it = queues_.lower_bound(front_priority_level_);
for (; it != queues_.end(); ++it) {
  ...
}

Comment (QMassoz) [reply src/scheduler_utils.cc:319]: Or we can drop the use of "front_priority_level_", as the request to be dequeued is likely to be found in the first key values of "queues_" (excepted when there is a large amount of user-specified non-default PolicyQueue)

Comment (dyastremsky) [reply src/scheduler_utils.cc:319]: Thanks for explaining and addressing the other loop. This looks good to me.

I'm less familiar with these files, so I'll wait for @GuanLuo to give this PR a once-over as well,

**Comment by QMassoz** [comment]
Any update?

**Comment by QMassoz** [comment]
I pushed commit [003a2c1 ](https://github.com/triton-inference-server/core/pull/116/commits/003a2c16235daaf43edc1edbd38ebadceb9f2e2a) because of a segfault that occurred very rarely in very particular situations since updating to newer version of triton server/client.

**Comment by dyastremsky** [comment]
Thanks! Guan, are you able to review this PR?

File: src/scheduler_utils.cc
Comment (GuanLuo) [reply src/scheduler_utils.cc:320]: Let's keep `front_priority_level_` for now to minimize change until necessary, but I think `last_priority_level_` can be safely removed as it is no longer used anywhere.

**Review by GuanLuo — APPROVED** [review]
Thanks for the PR, it looks good to me. @dyastremsky do you want to trigger a pipeline for the change?

**Comment by dyastremsky** [comment]
Fantastic, sure. Running tests. Will approve and merge if it passes.

I can open a PR to remove last_priority_level_ once merged.

**Comment by dyastremsky** [comment]
Thank you for your contribution! Great work, merged.

**Pull request merged**
